### ISSUE TITLE
SIP004 - Support for AEADs implemented by large libraries

### ISSUE THREAD

**Issue by Mygod** [shadowsocks/shadowsocks-org#30]
Use AEADs to replace stream cipher + OTA. Previous discussion: #29.

Proposed AEAD algorithms:

* [x] ChaCha20-Poly1305 (see also: [xSocks](https://github.com/lparam/xSocks))
* [x] XChaCha20-Poly1305
* [ ] Salsa20-Poly1305
* [x] AES-256-GCM (faster but not low-end-box-friendly)

Update: The following shows an example of TCP stream in chacha20-ietf-poly1305 mode (original idea by @breakwa11 and @Noisyfox). Other AEAD should follow the similar format.

```
Cipher: chacha20-ietf-poly1305

TCP request (after encryption, *ciphertext*)
+--------+----------------+--------------+--------------+---------------+
| NONCE  | PayloadLen_TAG | *PayloadLen* | Payload_TAG  |   *Payload*   |
+--------+----------------+--------------+--------------+---------------+
|  12    |       16       |       2      |     16       |    Variable   |
+--------+----------------+--------------+--------------+---------------+

TCP Chunk (after encryption, *ciphertext*)
+--------------+------------+-----------+----------+
| DATA_LEN_TAG | *DATA_LEN* |  DATA_TAG |  *DATA*  |
+--------------+------------+-----------+----------+
|      16      |     2      |     16    | Variable |
+--------------+------------+-----------+----------+
```

**Comment by madeye** [comment]
My plan is forking from ourselves, dropping all the support of stream ciphers and OTA.

After that the we can provide a new version of the protocol (v2?).

**Comment by wongsyrone** [comment]
I want to clarify AES-GCM is the fastest on machines with particular instruction sets, for example, server and client are both running on modern Intel CPU.

**Comment by wongsyrone** [comment]
I agree with https://github.com/shadowsocks/shadowsocks-org/issues/30#issuecomment-272088373

**Comment by Mygod** [comment]
I don't see the reason to drop compatibility for stream ciphers for now as @madeye has said:

> However, I don't think any known attacks to shadowsocks protocol is realistic. And please don't forget that most of the traffic over shadowsocks is TLS...

People who has higher security standards could migrate to using these algorithms. And if benchmark shows its influence to performance is negligible we should start to encourage everyone to migrate to AEADs.

And I don't think adding these additional algorithms would be a lot of work (we need at most 1 more abstraction layer).

**Comment by wongsyrone** [comment]
They can coexist as different projects, it is easy to maintain IMO.

**Comment by Mygod** [comment]
Considering SIP003, I think it's easier to maintain if implemented in one single project.

**Comment by madeye** [comment]
To support AEADs, first we need to define chunks for the TCP stream. Here is a proposal:
```
+----------+----------+----------+----------+
| DATA.LEN |   NONCE  |    TAG   |   DATA   | ...
+----------+----------+----------+----------+
|     2    |    8     |    16    | Variable | ...
+----------+----------+----------+----------| 
```

Assuming average data length equals to a typical MTU - 26 (1492 - 26 = 1466), then the maximum transfer efficiency is 1466 / 1492 = 98%.

**Comment by Noisyfox** [comment]
So which part of the chunk is encrypted? DATA? Or the whole chunk? Or both of them?

**Comment by madeye** [comment]
@Noisyfox DATA.LEN, NONCE and TAG should not be encrypted. Only DATA is encrypted. TAG is generated from encrypted DATA. For more details: https://github.com/lparam/xSocks/blob/master/src/crypto.c#L25

**Comment by Noisyfox** [comment]
That's basically the pattern of TLS record if I did understand correctly.

**Comment by madeye** [comment]
@Noisyfox Yes, it's exactly.

**Comment by wongsyrone** [comment]
xsocks' looks like https://github.com/jedisct1/libsodium/blob/master/src/libsodium/crypto_secretbox/xsalsa20poly1305/ref/secretbox_xsalsa20poly1305.c

It is AE not AEAD.

**Comment by Mygod** [comment]
Is DATA.LEN necessary?

**Comment by madeye** [comment]
@Mygod Yes, I think so. It's required to split TCP stream to chunks.

**Comment by wongsyrone** [comment]
AE:

- Encrypts a message with a key and a nonce to keep it confidential
- Computes an authentication tag. This tag is used to make sure that the message hasn't been tampered with before decrypting it.

AEAD:

- Encrypts a message with a key and a nonce to keep it confidential
- Computes an authentication tag. This tag is used to make sure that the message, **as well as optional, non-confidential (non-encrypted) data**, haven't been tampered with.

AEAD can protect plain text as well.

**Comment by Mygod** [comment]
@madeye What about letting it equal to packet length - 24?

@wongsyrone I don't think we have non-confidential data to protect however.

**Comment by madeye** [comment]
@Mygod In a TCP stream, there is not packet length...

**Comment by Mygod** [comment]
Oh sorry I got confused because you mentioned MTU. 😅

**Comment by madeye** [comment]
MTU here is only to demonstrate how large is a chunk in the real world... In our implementation, we should also limit the DATA.LEN, make it smaller than the socket buffer size, e.g. 2048 in shadowsocks-libev.

**Comment by wongsyrone** [comment]
NONCE is too short.

> Note: the currently included AEAD cipher based on the draft only uses an eight byte nonce, which is too short to be safely generated randomly. A later version of libsodium will include the updated version with a longer nonce.

https://crypto.stackexchange.com/questions/29311/how-do-i-tack-on-additional-data-to-libsodiums-public-key-authenticated-encry

**Comment by madeye** [comment]
Then what about this:

```
+----------+----------+----------+----------+----------|
| DATA.LEN |   NONCE  |   SEQ    |    TAG   |   DATA   | ...
+----------+----------+----------+----------+----------+
|     2    |    8     |    8     |    16    | Variable | ...
+----------+----------+----------+----------+----------+
```

SEQ is a 64-bit sequence number. The real NONCE is NONCE + SEQ. SEQ is also the AD of the AEAD.

(Why shouldn't I use TLS directly? 😅 )

**Comment by wongsyrone** [comment]
[draft] main idea from @Noisyfox 
The purpose of Length tag is to keep length as-is and not tempered by the third party since this is the key to separate general ciphertext and auth tag.

Please note that this is not compatible with current OTA and original protocol.

```c
/*
 * The main difference between OTA designed by madeye and this one is MtE vs EtM.
 *
 * The first nonce is either from client or server side, it is generated via randombytes_buf()
 * in libsodium, and the sequent nonces are incremented via sodium_increment() in libsodium.
 * IMPORTANT: nonce should be used only once, let us running on the right track.
 *
 * Data.Len is used to seperate general ciphertext and Auth tag. We can start decryption
 * iff passing verification.
 *
 * For UDP, nonces are generated randomly without the incrementation.
 *
 * TCP request (before encryption)
 * +------+---------------------+------------------+
 * | ATYP | Destination Address | Destination Port |
 * +------+---------------------+------------------+
 * |  1   |       Variable      |         2        |
 * +------+---------------------+------------------+
 *
 * TCP request (after encryption)
 * +--------+-----------+---------------+----------+---------------+
 * | NONCE  |PayloadLen |PayloadLen_TAG | Payload  |  Payload_TAG  |
 * +--------+-----------+---------------+----------+---------------+
 * | Fixed  |     2     |     Fixed     | Variable |     Fixed     |
 * +--------+-----------+---------------+----------+---------------+
 * 
 * Payload input: atyp + dst.addr + dst.port
 * PayloadLen is length(atyp + dst.addr + dst.port)
 * Payload_TAG and PayloadLen_TAG are in plaintext
 *
 * TCP Chunk (before encryption)
 * +----------+
 * |  DATA    |
 * +----------+
 * | Variable |
 * +----------+
 *
 * Data.Len is a 16-bit big-endian integer indicating the length of DATA.
 *
 * TCP Chunk (after encryption)
 * +------------+---------+-----------+----------+
 * | Data.Len*  | Len_TAG |  DATA_TAG |  DATA*   |
 * +------------+---------+-----------+----------+
 * |      2     | Fixed   |  Fixed    | Variable |
 * +------------+---------+-----------+----------+
 *
 * Len_TAG and DATA_TAG have the same length, they are in plaintext.
 * After encryption, DATA -> DATA*
 *
 * UDP (before encryption)
 * +------+---------------------+------------------+----------+
 * | ATYP | Destination Address | Destination Port |   DATA   |
 * +------+---------------------+------------------+----------+
 * |  1   |       Variable      |         2        | Variable |
 * +------+---------------------+------------------+----------+
 *
 * UDP (after encryption)
 * +--------+----------+-----------+
 * | NONCE  |  DATA*   |  DATA_TAG |
 * +--------+----------+-----------+
 * | Fixed  | Variable |  Fixed    |
 * +--------+----------+-----------+
 *
 * DATA* is Encrypt(atyp + dst.addr + dst.port + DATA)
 * RSV and FRAG are dropped
 * Since UDP packet is either received completely or missed,
 * we don't have to keep a field to track its length.
 *
 */
```

**Comment by v2ray** [comment]
I don't see the point of "length-tag". AEAD support additional data to be authenticated as plaintext. You can use length as the "additional data" for temper proofing.

**Comment by madeye** [comment]
@v2ray Agreed. The DATA.LEN can be protected by AEAD as plaintext.

**Comment by Mygod** [comment]
+1. But this reminds me that we need random noise indistinguishability more than authentication since the original purpose of this repo is to circumvent GFW. Let's review our additions against random noise indistinguishability:

1. TAG: Check. This should be guaranteed by the security of the hash algorithm we used.
2. DATA.LEN: Oh shit this is not random at all. Adversary can distinguish them very easily.
3. NONCE: No problem as long as they are randomly generated. (which isn't the case in TCP)
4. SEQ: Terrible idea. It makes distinction much more easier.

OTA also has this problem. So what do you think?

**Comment by Mygod** [comment]
My thoughts: it's theoretically feasible to achieve both indistinguishabilities. Here's a possible solution:

Encryption input: 
1. Data (256 bytes)
2. Pad (1 byte): how many bytes (0-255) should be dropped from data.

AEAD unit:
1. Tag (16 bytes)
2. Encryption output (257 bytes)

Nonce is sent in the initial packet and auto incremented every x units (it should be small enough to prevent a replay attack) and/or when pad != 0 (this one sounds more reasonable).

The problem with this approach is that it can be easily distinguished by looking at its length (it's always a multiple of 273). So maybe whenever pad != 0, a random noise of length = pad should be appended.

**Comment by Mygod** [comment]
Limitations of the above approach: redundant data used for authentication and indistinguishability takes 6% - 93% (although in large transfers it should always stay low)

Now that I think again, another less hacky approach is to send data.len with AEAD then send data with AEAD. :P

**Comment by Mygod** [comment]
*removes some of my useless comments*

Okay I think I got this sorted out. What we need to do is just to put data.len in the encryption data; send nonce ONLY in the first data and it auto increments for EVERY data. And voilà!

**Comment by v3aqb** [comment]
FYI: http://github.com/v3aqb/hxsocks

Mygod <notifications@github.com>于2017年1月18日周三 22:52写道：

**Comment by Mygod** [comment]
@v3aqb

> better encrypted

I beg to differ. I took a quick glance and you implemented what shadowsocks was designed to prevent.

**Comment by madeye** [comment]
@Mygod Then the only difference from our current OTA is that the new proposal is Encrypt-Then-MAC... :smile:

**Comment by wongsyrone** [comment]
This is just an example of improving content security (aka CCA-proof), while it undermines protocol security (aka censorship-bypassing-detection-proof).

Before any progress, we should examine it carefully and make the right choice.

**Comment by Mygod** [comment]
@madeye No. There are two significant differences:

1. Data.len is transmitted in ciphertext instead of plain text. This implies that it's easy for GFW to detect shadowsocks connections with OTA enabled.
2. AE algorithms are chosen and implemented by large library and tested/attempted attack by more people.

@wongsyrone Yes. We need to know which one is more important in this context. I think my approach should be safest afaic.

**Comment by v2ray** [comment]
@Mygod if data.len is encrypted, how to figure out the length of chunk for decryption (in the context of AEAD)?

**Comment by wongsyrone** [comment]
The noise data is another approach to obfuscating packet length pattern, which may help.

Without data.len authentication, you cannot assume it hasn't been touched. That's why we decide to add a tag for it.

Encryption never implies authentication.

In fact, the AD part is never used in the proposal; I planned to make it work like another pre-shared info and verify it before encryption. It looks like a kind of overkill.

**Comment by Noisyfox** [comment]
@v2ray You get the point man.

**Comment by Mygod** [comment]
@v2ray Just decrypt the first 2 bytes for stream cipher or first block for block cipher. I don't see a problem.

@wongsyrone All encryption I mentioned are authenticated encryption. data.len will be included when calculating hash as well.

**Comment by Noisyfox** [comment]
@Mygod If some one increase the length by one byte, then the server will waiting for one more byte instead of close the connection immediately.

**Comment by Noisyfox** [comment]
Well you see, AEAD's main point is never decrypt any data before the cipher text is been verified.

**Comment by Mygod** [comment]
@Noisyfox Good point. So maybe we should authenticate data.len and data separately instead? Like `TAG`, `encrypted_data.len`, `TAG`, `encrypted_data`?

**Comment by Mygod** [comment]
@Noisyfox Hmm on second thought actually that may not be a big issue. If a man-in-the-middle removes the last byte from the original stream, the server is also going to wait for that last byte to arrive instead of closing the connection immediately. A malformed data.len isn't a big issue IMO. What do you think?

**Comment by wongsyrone** [comment]
The proposal is supposed to encrypt length and corresponding data separately.
The length verification laies on first class. The second is data verification.

**Comment by Noisyfox** [comment]
@Mygod But if modify one byte in cipher text will lead to server's blocking, then the attacker could find out that one specific byte means length. Well I think it's better to hide our protocol too.

**Comment by Mygod** [comment]
@wongsyrone Oh you're right. I didn't see that there was an asteroid after `data.len` which means encrypted. Sorry about that y'all. 😛

@Noisyfox Yes. That's a better attack! Now I'm convinced that we should take your proposal.

**Comment by v2ray** [comment]
There is a simpler way for 'encrypting' data.len. Only the first 'data.len' matters. If the attacker can't figure out where is (or what is) the first 'data.len', he/she can't crack the following data.len as well. To simplify the design (and also increase traffic throughput), you can just encrypt the data.len for the first chunk of payload.

**Comment by Mygod** [comment]
You ignored timing. TCP stream can pause for some time and then resumed. Also we should never put plain text in the traffic unless it's indistinguishable with random noise.

**Comment by wongsyrone** [comment]
Just reference, we may borrow some ideas from https://github.com/WeMobileDev/article/blob/master/%E5%9F%BA%E4%BA%8ETLS1.3%E7%9A%84%E5%BE%AE%E4%BF%A1%E5%AE%89%E5%85%A8%E9%80%9A%E4%BF%A1%E5%8D%8F%E8%AE%AEmmtls%E4%BB%8B%E7%BB%8D.md

**Comment by madeye** [comment]
Great! I think we're very close to the new protocol.

@Mygod Could you add @Noisyfox's design to your proposal?

**Comment by wongsyrone** [comment]
Updated draft to describe the procedure.

**Comment by Mygod** [comment]
Maybe put tag before encrypted data.len since "never decrypt any data before the cipher text is been verified"?

**Comment by wongsyrone** [comment]
In general, there are three parts after encryption:

1. Nonce
2. data.len and its tag
3. data and its tag

The sequence of tag and ciphertext in each part doesn't matter since tags are sharing the same length.
The difference is how to compute the offset.

**Comment by Mygod** [comment]
I meant TAG, data.len, TAG, data. It just looks nicer.

**Comment by wongsyrone** [comment]
I see, no objection from my side.

**Comment by wongsyrone** [comment]
Just come up with another idea, How about dropping data tag, use only

> data.len_TAG, data.len, data

**Comment by Mygod** [comment]
Then this is almost no different with the current protocol without OTA.

**Comment by wongsyrone** [comment]
At least, it ensures datalen is not compromised. It's just a hint.

**Comment by madeye** [comment]
Another problem is about data.len_TAG. Should we use the full length of AEAD for two bytes of data.len?

**Comment by Mygod** [comment]
@wongsyrone Also your approach is also vulnerable to the detect method proposed by @breakwa11 as long as you are using stream cipher.

@madeye 2 bytes may be too short. Only 65,536 combinations for attacker to try.

**Comment by wongsyrone** [comment]
The sad truth is crypto library expects a full-length tag.

**Comment by Mygod** [comment]
...and attaching full length tag isn't a big overhead IMO.

**Comment by wongsyrone** [comment]
Another problem is key derivation, I proposed to use crypto_generichash() in libsodium, which uses Blake2b internally. Or some real password hashing algorithms like Argon2 in libsodium.

**Comment by wongsyrone** [comment]
> 2 bytes may be too short. Only 65,536 combinations for attacker to try.

When computing the tag, we can add user input as AD. No matter how many times you tried, you don't know AD; then it will result in auth failure.
After auth failure, close connection instantly or add them to a list with a random timeout, close conn when timeout exceeds.

**Comment by Mygod** [comment]
@wongsyrone Good points except I don't see the point of using AD. I assume your user input is some kind of pre-shared message. In that case we already have a secret key.

**Comment by wongsyrone** [comment]
Yes, it is pre-shared message work like password we already have, we can reuse it or add a similar entry.

**Comment by Mygod** [comment]
I still don't think it's necessary.

**Comment by breakwa11** [comment]
各位官爷，本女才疏学浅，AT在下有何贵干

**Comment by madeye** [comment]
@wongsyrone As we don't have any plaintext now (except the nonce and tag), I think there is no need to use AD here.

I'd like to summarize the ideas above. Now, we have a proposal like this:

```
Cipher: chacha20-poly1305

TCP request (after encryption, *ciphertext*)
+--------+----------------+--------------+--------------+---------------+
| NONCE  | PayloadLen_TAG | *PayloadLen* | Payload_TAG  |   *Payload*   |
+--------+----------------+--------------+--------------+---------------+
|  12    |       32       |       2      |     32       |    Variable   |
+--------+----------------+--------------+--------------+---------------+

TCP Chunk (after encryption, *ciphertext*)
+--------------+------------+-----------+----------+
| DATA_LEN_TAG | *DATA_LEN* |  DATA_TAG |  *DATA*  |
+--------------+------------+-----------+----------+
|      32      |     2      |     32    | Variable |
+--------------+------------+-----------+----------+
```

**Comment by wongsyrone** [comment]
It should be:

```
Cipher: chacha20-ietf-poly1305

TCP request (after encryption, *ciphertext*)
+--------+----------------+--------------+--------------+---------------+
| NONCE  | PayloadLen_TAG | *PayloadLen* | Payload_TAG  |   *Payload*   |
+--------+----------------+--------------+--------------+---------------+
|  12    |       16       |       2      |     16       |    Variable   |
+--------+----------------+--------------+--------------+---------------+

TCP Chunk (after encryption, *ciphertext*)
+--------------+------------+-----------+----------+
| DATA_LEN_TAG | *DATA_LEN* |  DATA_TAG |  *DATA*  |
+--------------+------------+-----------+----------+
|      16      |     2      |     16    | Variable |
+--------------+------------+-----------+----------+
```

**Comment by wongsyrone** [comment]
Non-ietf variant, change Nonce from 12 to 8.
For xchacha20-ietf-poly1305, change Nonce from 12 to 24.

**Comment by wongsyrone** [comment]
Non-IETF versions of crypto_aead_xchacha20poly1305 is already removed: https://github.com/jedisct1/libsodium/commit/1686da3d3cf1d3c40a6985f8c515d20feb451bf8

BTW, xchacha20-ietf-poly1305 is not released yet.

For shadowsocks-libev:

```c
#ifdef crypto_aead_xchacha20poly1305_ietf_ABYTES
#define HAVE_XCHACHA20IETF
#endif

static const char *supported_ciphers[CIPHER_NUM] = {
    "aes-128-gcm",
    "aes-192-gcm",
    "aes-256-gcm",
    "chacha20-poly1305",
    "chacha20-poly1305-ietf",
#ifdef HAVE_XCHACHA20IETF
    "xchacha20-poly1305-ietf"
#endif
};

/*
 * use mbed TLS cipher wrapper to unify handling
 */
#ifdef USE_CRYPTO_MBEDTLS
static const char *supported_ciphers_mbedtls[CIPHER_NUM] = {
    "AES-128-GCM",
    "AES-192-GCM",
    "AES-256-GCM",
    CIPHER_UNSUPPORTED,
    CIPHER_UNSUPPORTED,
#ifdef HAVE_XCHACHA20IETF
    CIPHER_UNSUPPORTED
#endif
};
#endif

static const int supported_ciphers_nonce_size[CIPHER_NUM] = {
    12, 12, 12, 8, 12,
#ifdef HAVE_XCHACHA20IETF
    24
#endif
};

static const int supported_ciphers_key_size[CIPHER_NUM] = {
    16, 24, 32, 32, 32,
#ifdef HAVE_XCHACHA20IETF
    32
#endif
};

static const int supported_ciphers_tag_size[CIPHER_NUM] = {
    16, 16, 16, 16, 16,
#ifdef HAVE_XCHACHA20IETF
    16
#endif
};
```

**Comment by breakwa11** [comment]
> Why shouldn't I use TLS directly?

so, why?

**Comment by Mygod** [comment]
@madeye I think plugins are basically settled for now (unless one's starting to port PT to our codebase). Let's start implementing this because I think it's pretty urgent. A man-in-the-middle should be able to easily identify OTA-enabled trafficc since there is only 255 possibilities to try. I was trying to write a plugin based on simple-obfs to serve as a emulated man-in-the-middle to illustrate my attack but I haven't finished debugging it yet. After implementing this we should deprecate OTA.

**Comment by madeye** [comment]
@Mygod Could you update the final frame design in your proposal? I'm OK with the design of @wongsyrone. So let's just follow it https://github.com/shadowsocks/shadowsocks-org/issues/30#issuecomment-273692801.

**Comment by v2ray** [comment]
@Mygod where does the number 255 come from? The address type? If so, are you saying the new GCM method also applies to the header part?

**Comment by Mygod** [comment]
@madeye Done. (btw I believe you can edit my comment directly)

@v2ray Address length varies from 2 bytes to 256 bytes inclusive. The most obvious problem of the old approach is that its data.len is transmitted in plaintext.

**Comment by breakwa11** [comment]
You'd better mention my name if you choose to follow [this design](https://github.com/shadowsocks/shadowsocks-org/issues/30#issuecomment-273692801) which idea was came from me actually

**Comment by madeye** [comment]
@breakwa11 Yep, I think @Mygod has already added your name to his proposal description. https://github.com/shadowsocks/shadowsocks-org/issues/30#issue-200289203

**Comment by riobard** [comment]
Please reverse the order of TAG and PAYLOAD (so it's PAYLOAD then TAG), as it's easier to use the normal combined mode AEAD encrypt/decrypt functions instead of the detached mode (e.g. in libsodium). Similarly, Go's standard library has a fixed AEAD signature assuming the normal combined mode order (https://golang.org/pkg/crypto/cipher/#AEAD). 

Since the whole point of using proven AEAD is to prevent developers from screwing up crypto, it is preferable to follow conventions to reduce complexity.

**Comment by Mygod** [comment]
Sounds good to me.

**Comment by madeye** [comment]
LGTM. Although I don't think the order here really matters, golang indeed implement GCM in that way: https://golang.org/src/crypto/cipher/gcm.go#147

**Comment by wongsyrone** [comment]
It doesn't matter.

**Comment by wongsyrone** [comment]
Updated nonce incrementation:
> Please note that sodium_increment() considers the number to be encoded in little-endian format.

**Comment by riobard** [comment]
For key derivation, libsodium's `crypto_generichash` is marginally better than MD5/SHA1. 

It's safer to use proven password hashes such as PBKDF2. Argon2 is fine, but I worry about lack of mature library support due to its novelty. Additionally, we have to choose the right parameters (number of iterations, salt, etc) for password hashes since most of them require knobs…

**Comment by madeye** [comment]
Here is the experimental branch for SIP004 https://github.com/shadowsocks/shadowsocks-libev/tree/sip004

Please help to run tests and refine the document by adding comments here.

**Comment by wongsyrone** [comment]
The password hashing looks overkill, I prefer to generic hashing like blake2b.

**Comment by madeye** [comment]
@wongsyrone Let's keep using previous key driviation (OpenSSL way). Not every shadowsocks port is able to use blake2b hashing.

**Comment by em0minate** [comment]
I'd suggest at least upgrade to `PBKDF2` for key derivation.

**Comment by wongsyrone** [comment]
Just wondering which port? According to its official website, there are many implementations available, and many libsodium binding wrappers.

**Comment by madeye** [comment]
@wongsyrone qt5, golang, nodejs, chromeapp, rust. Also, if we are really care about password hashing, I prefer argon2 here, as it's a memory-hard approach.

**Comment by wongsyrone** [comment]
I think we should get rid of MD5 at least. Argon2 is good alternative as well.

**Comment by riobard** [comment]
If we go for Argon2, we need to choose which variant of Argon2 as there are three: Argon2i, Argon2d, and Argon2id. According to https://github.com/P-H-C/phc-winner-argon2, Argon2i seems to be a better fit in our case

> Argon2 has three variants: Argon2i, Argon2d, and Argon2id. Argon2d is faster and uses data-depending memory access, which makes it highly resistant against GPU cracking attacks and suitable for applications with no threats from side-channel timing attacks (eg. cryptocurrencies). *Argon2i instead uses data-independent memory access, which is preferred for password hashing and password-based key derivation, but it is slower as it makes more passes over the memory to protect from tradeoff attacks*. Argon2id is a hybrid of Argon2i and Argon2d, using a combination of data-depending and data-independent memory accesses, which gives some of Argon2i's resistance to side-channel cache timing attacks and much of Argon2d's resistance to GPU cracking attacks.

**Comment by riobard** [comment]
Alternatively, we could leave the decision of which KDF to use to deployment and require a strong key instead.

**Comment by wongsyrone** [comment]
It seems things are getting complicated now. Since we are not passing it directly, does it really matter to use real password hashing algorithm?

**Comment by riobard** [comment]
> Since we are not passing it directly, does it really matter to use real password hashing algorithm?

Yes if you want the server to be resistant to password guessing attacks. 

No if we move it out of the scope of this project and require long random keys.

**Comment by wongsyrone** [comment]
The reason why I prefer to blake2b is it claims as secure as SHA3 but slightly faster than MD5 and SHA1. Blake2b is also used in Argon2.

**Comment by riobard** [comment]
blake2b, SHA, and any fast hashing in general are not suitable for password hashing. 

You need slower, tunable password hashing like PBKDF2, bcrypt, scrypt, or Argon2 (each has its own crypto tradeoffs).

**Comment by riobard** [comment]
Ultimately, though, nothing beats long random keys from `/dev/urand` ^_^

**Comment by wongsyrone** [comment]
Okay, I leave this to madeye to balance.

**Comment by em0minate** [comment]
Furthermore, isn't it necessary a configuration field for hashing salt as well?

**Comment by Mygod** [comment]
For maximum security, we can also get rid of key derivation by using a random generated key file (or base64 encoded string) like PGP.  This also guarantees compatibility on every platform ever.

**Comment by madeye** [comment]
Let's use Argon2 as our default key derivation for AEAD (nothing changed for old stream ciphers).

We follow the predefined settings from libsodium and take "shadowsocks hash" as the default salt.

```
        const unsigned char salt[crypto_pwhash_SALTBYTES] = {
            's', 'h', 'a', 'd', 'o', 'w', 's', 'o',
            'c', 'k', 's', ' ', 'h', 'a', 's', 'h'
        };
        int err = crypto_pwhash (key, nkey, (char*)pass, strlen(pass), salt,
                crypto_pwhash_OPSLIMIT_INTERACTIVE, crypto_pwhash_MEMLIMIT_INTERACTIVE,
                crypto_pwhash_ALG_DEFAULT);
```

More details: 

1. https://github.com/shadowsocks/shadowsocks-libev/commit/e18f71b03fb54a0d34b6aa41f8966620722c4c56
2. https://download.libsodium.org/doc/password_hashing/the_argon2i_function.html

**Comment by riobard** [comment]
The above Argon2 settings require 4MB memory to compute keys. Not sure if it's gonna be ok on low end routers.

**Comment by madeye** [comment]
@riobard 4MB memory is not a problem for current low end routers. (32MB DDR2, MT7620A, $10).

**Comment by riobard** [comment]
Great!

**Comment by riobard** [comment]
Do we have a limit for PayloadLen? Currently the field is an unsigned 16-bit integer so maximum possible length is 0xFFFF = 65,535 (64KB - 1). For comparison, TLS record payload is limited to 16KB.

For Shadowsocks implementations using either AEAD or OTA, PayloadLen impacts throughput and latency (as measured in TTFDB = Time To First Decrypted Byte):

- Lower PayloadLen incurs more fragmentation of the original data stream, thus more framing overhead (currently about 34 bytes per record/chunk) and lower throughput;
- Higher PayloadLen increases TTFDB as clients need to wait for all payload packets to arrive before decryption and authentication can complete.

For example, assuming a typical TCP packet size of 1450, a 64KB payload will split into over 45 packets. If any of those packets is lost during transmission, we have to wait for re-transmission. Worse, since most Shadowsocks deployments are over high-latency high-loss cross-border links, the delay introduced due to re-transmission will be amplified.

16KB seems to be reasonable number for typical uses, but even in that case there're discussions about dynamically sizing TLS records to reduce TTFDB. For more information please read https://www.igvita.com/2013/10/24/optimizing-tls-record-size-and-buffering-latency/

I'd propose we limit PayloadLen to 0x3FFF = 16,383 (16KB - 1): that's the lower 14 bits of the big-endian encoded 16-bit field, leaving the higher 2 bits reserved for future use (e.g. flags for more payload types?). The proposed chunk structure would be

```
[2-bit *reserved (must be 00)*][14-bit *payload_len*][payload_len tag][payload][payload tag]
```

Implementations are free to dynamically size chunks similar to the optimization utilized by some TLS implementations.

**Comment by riobard** [comment]
@madeye I was wrong about the memory requirement of Argon2. Actually `crypto_pwhash_MEMLIMIT_INTERACTIVE` requires 32 megabytes of memory. I was misled by the "Mb" notation in libsodium's doc which says

> For interactive, online operations, crypto_pwhash_OPSLIMIT_INTERACTIVE and crypto_pwhash_MEMLIMIT_INTERACTIVE provide base line for these two parameters. This requires 32 Mb of dedicated RAM. Higher values may improve security (see below).

Checking the header definition here https://github.com/jedisct1/libsodium/blob/68564326e1e9dc57ef03746f85734232d20ca6fb/src/libsodium/include/sodium/crypto_pwhash_argon2i.h#L35 reveals it's actually 32 MB of memory, not Mb.

**Comment by riobard** [comment]
Come to think about it, memory-hard password hashing is not suitable for low-end routers with limited memory. We could lower the memory requirement to 4MB, but then it becomes pointless to use memory-hard hashing in the first place.

I'm now inclined to either a) leave this decision out of the scope and require long random keys instead; or b) use PBKDF2.

**Comment by v2ray** [comment]
> a) leave this decision out of the scope and require long random keys instead;

One step closer to [VMess protocol](https://www.v2ray.com/chapter_04/03_vmess.html). How about adding an one time auth at the beginning of the request then?

**Comment by madeye** [comment]
@riobard 4MB (or 32MB) is already large enough for current hardware. Imagine a typical GPU with 1024 CUDA cores and 4GB memory, it's totally impossible to implement efficient parallel password hashing algorithm on it. (Usually we need more than 1024 threads to hide the DRAM latency, but over 1GB memory usage would already cause huge L2, or even TLB cache misses).

We shouldn't worry about low end routers, as they can keep using stream ciphers. Actually, AEAD is not friendly to them: Even with chacha20/rc4 stream cipher, MT7620a could only achieve about 20Mbps bandwidth. If a user wants better perf and security, they should upgrade their routers first. 

The largest possible payload length in shadowsocks protocol is around 256 bytes, as there is a limit of domain name length. Typically the first payload (shadowsocks request header) will be sent with several subsequent chunks (usually HTTP/TLS header) in one packet.

**Comment by wongsyrone** [comment]
Since the salt is pre-defined, it is pointless to use password hashing, it requires salt to be randomly generated.

I think we get too far here, generic hashing is enough.

**Comment by wongsyrone** [comment]
> In cryptography, a salt is **random data** that is used as an additional input to a one-way function that "hashes" a password or passphrase.

https://en.wikipedia.org/wiki/Salt_(cryptography)

**Comment by madeye** [comment]
@wongsyrone It's a just a default salt, you're free to change it.

**Comment by wongsyrone** [comment]
We want to compute the same key on both client and server side; password hashing will drop to generic hashing. For Argon2i, it will generate the same output if and only if with the same human-written password, salt, parameters. It is equivalent to ordinary hashing.

Although you can change the default salt, you have to keep both sides using the same salt.

**Comment by madeye** [comment]
@wongsyrone Actually not. Password hashing and generic hashing are totally different. Please find more details in @riobard's comments:

> blake2b, SHA, and any fast hashing in general are not suitable for password hashing.
>
> You need slower, tunable password hashing like PBKDF2, bcrypt, scrypt, or Argon2 (each has its own crypto tradeoffs).

**Comment by wongsyrone** [comment]
Yes, they are indeed not suitable for password hashing, but there is an assumption: do you want to store the password and protect it from being compromised? We are not storing them, in fact, we just use it to encrypt traffic. Password are in plaintext stored in JSON config file.

**Comment by wongsyrone** [comment]
If I can get your entire JSON config file, no matter what kind of password hashing is being used, I can still connect to your server.

**Comment by madeye** [comment]
@wongsyrone Imagine this situation, one man decrypts your shadowsocks traffic with brute-force  approach in the future (or a defect of a stream cipher). Even if he gets the key, it's almost impossible to get the password with a good password hashing.

Hide your key file is not the responsibility of shadowsocks.

**Comment by wongsyrone** [comment]
> one man decrypts your shadowsocks traffic with brute-force approach 

That's why we need AE; brute-force approach won't work since auth failure.

> a defect of a stream cipher

also the reason to equip AE

---

In fact, someone doesn't have to know your password; it is pointless, he only wants to know whether you are bypassing censorship, then he can block your IP address. Done.

**Comment by wongsyrone** [comment]
If you can find a way to crack the current AEAD design, we can simply drop this design now.

**Comment by madeye** [comment]
Authentication cannot avoid any brute-force attack to the encrypted traffic. For example,

1. Capture and store all the traffic to/from your shadowsocks server.
2. With enough computation power (or after several years), it's possible to decrypt your traffic by brute-force the key.


AE in this proposal is to provide IND-CCA2, which has nothing to do with stream cipher itself.

**Comment by wongsyrone** [comment]
Only if you are a valuable target. Someone will be eager to have a cup of tea with you and monitor your behavior all day. At that time, Shadowsocks won't work anymore.

**Comment by madeye** [comment]
@wongsyrone Actually, it's not an expensive attack in terms of computation power or storage today. To improve the efficiency, the attacker can just store the first packet, as it's basically atyp + domain name.

The point here is if we have better approach today, why shouldn't we use it?

**Comment by riobard** [comment]
@madeye I'm all good if we can just ignore low-end routers for the new design. Currently I'm using an x86 router with AES-NI. AES-GCM performs really well on that fanless little box :P

> And the sizes of subsequent chunks are actually limited by the socket buffer size, as we send every packet (equal or smaller than the socket buffer size of an implementation) we received in a chunk.

Most implementations will indeed work that way, but nothing stops an implementation from batching up chunks. At this stage of protocol design, it is wise to put a cap on the payload length just in case.

To be honest, my secret hope is to keep those 2 reserved bits for future experiments (how about some alert and control messages)! ^_^

**Comment by madeye** [comment]
> To be honest, my secret hope is to keep those 2 reserved bits for future experiments (how about some alert and control messages)! ^_^

It looks a good point. I'm OK with adding this limit.

**Comment by riobard** [comment]
@v2ray I've read thru the VMess doc and it's a way more complex protocol. What's the purpose of the initial OTA?

**Comment by madeye** [comment]
Added `0x3FFF` as the chunk size mask and limit. https://github.com/shadowsocks/shadowsocks-libev/commit/c89f5f3f3f93ca3c671ce1a61ea55b5563226136

**Comment by v2ray** [comment]
@riobard For fast authentication. 1) The authentication part is always 16 bytes long. No need to parse variable length of input to validate the request. 2) Low cost on server. It only cost server an O(1) dict-loopup for authentication.

**Comment by riobard** [comment]
@v2ray Thanks for the explanation! If I understand correctly, I don't think it applies here as there's no user auth at all. Basically we're designing a framing protocol. Application-level concern can be left to upper layer.

**Comment by v2ray** [comment]
@riobard It is related as you are discussing password generation and somewhat "AEAD" above. User authentication in Shadowsocks is implied by request parsing: if request is parsed successfully, the request is authenticated, and vice versa. The implied authentication triggers some attacks. That's why the DATALEN_TAG is necessary, as discussed above.

Since now the topic shifts to password generation (and even randomly like TLS), it may be worth to go one step further to break apart the authentication and request parsing.

**Comment by riobard** [comment]
@v2ray I'm afraid you cannot cleanly separate the two in shared secret encryption, as "request parsing" implies decryption, which further implies authenticating the shared secret.

**Comment by v2ray** [comment]
@riobard Theoretically it is possible. Once the authentication is finished, the rest of the TCP stream can be encrypted (and decrypted) by another set of secrets. See [Zero-RTT handshakes in TLS 1.3](https://timtaubert.de/blog/2015/11/more-privacy-less-latency-improved-handshakes-in-tls-13/)

**Comment by riobard** [comment]
12-byte nonce is too short.

TLS can get by because it generates a random key for each connection, but we're using the same key for all connections and incrementing nonce for each chunk, nonce space will be exhausted very fast and Birthday Paradox will bite us.

Nonce size must be at least 128 bits to be on the safe side. Periodic key rotation is advised for additional safety.

I'd propose our cipher suites should follow TLS 1.3 recommendations (https://tlswg.github.io/tls13-spec/#rfc.section.8.1), which are:

1. AES-128-GCM (mandatory)
2. ChaCha20-Poly1305 (optional)
3. AES-256-GCM (optional)

Deployments should prefer AES-GCM for devices with AES-NI (most modern machines will do), and ChaCha20-Poly1305 for devices without.

**Comment by riobard** [comment]
@v2ray Not really. TLS 1.3 can do that because it reuses information from previous exchange (which was established by asymmetric encryption, not shared secret). Shadowsocks and VMess cannot do so because both use symmetric key encryption exclusively.

In fact, the encryption key transmitted in VMess is redundant—there's no benefit at all. You might as well use the randomly generated user ID as key. You might need to seriously re-consider the design as v1 shows many over-engineering symptoms.

**Comment by Mygod** [comment]
Okay I'm back (from playing a lot of Stardew Valley recently).

@wongsyrone AEAD can better prevent online brute-force (which should be more expensive than offline brute-force but can be used for server detection) but there's nothing AEAD can do about offline brute-force attacks.

@madeye Sounds like you have some kind of benchmark. Can you share some details?

@v2ray Authentication here refers to authenticating the integrity of the data. The authentication in your design refers to authenticating the user ID (or the pre-shared secret). Encrypting with a random secret would be more secure but it will introduce handshakes which Shadowsocks is designed to get rid of. In that case, just use SSH tunnel.

**Comment by madeye** [comment]
@Mygod No particular benchmark for Argon2. I just did some quick math to explain the reason why memory-hard hashing is difficult to parallel on our current hardware (Von Neumann architecture).

But we indeed have internal benchmarks for Ethash (PoW of Ethereum). For these kind of hashing algorithm, huge L2 cache misses and even TLB cache misses are expected.

**Comment by pexcn** [comment]
https://shadowsocks.org/en/spec/aead.html @madeye Page not found.

**Comment by Mygod** [comment]
I suppose we can close this now as this has been added to shadowsocks.org.

**Comment by riobard** [comment]
Nonce size is still an open question. I guess we could leave it up to deployment time?

**Comment by madeye** [comment]
@riobard Any suggestion? Which size is large enough?

**Comment by riobard** [comment]
As explained in https://github.com/shadowsocks/shadowsocks-org/issues/30#issuecomment-276073728, if my math is right, we need at least 128 bits. 224 or 256 is even better, given that the impact on performance is negligible. Nonce is sent only once in TCP. Impact on UDP traffic should also be acceptable if used for tunneling DNS only.

However this is non-essential to the protocol design, but rather a choice of cipher suite and its parameters. For example, we could make nonce size a command line parameter. I did that in my fresh Golang implementation (which also leaves out the choice of KDF and requires random keys).

**Comment by madeye** [comment]
@riobard I'm going to update https://github.com/shadowsocks/shadowsocks-org/blob/gh-pages/src/content/en/spec/02-cipher.md. What about this?

```
Method Name	Key Size	IV Length	Tag Length
aes-128-gcm	16	16	16
aes-192-gcm	24	24	16
aes-256-gcm	32	32	16
chacha20-poly1305	32	16	16
chacha20-ietf-poly1305	32	24	16
xchacha20-ietf-poly1305	32	32	16
```

**Comment by wongsyrone** [comment]
These are specified by ciphers, please don't do that.
https://tools.ietf.org/html/rfc7539#section-2.3

**Comment by madeye** [comment]
@wongsyrone Any document for aes-gcm?

**Comment by wongsyrone** [comment]
https://github.com/ARMmbed/mbedtls/blob/development/library/cipher_wrap.c#L369

**Comment by wongsyrone** [comment]
You need to check definition of mbedtls_cipher_info_t to know the meanings

**Comment by madeye** [comment]
OK. What about this?

<table style="width:100%">
  <tr>
    <th>Method Name</th>
    <th>Key Size</th>
    <th>IV Length</th>
    <th>Tag Length</th>
  </tr>
  <tr>
    <td>aes-128-gcm</td>
    <td>16</td>
    <td>12</td>
    <td>16</td>
  </tr>
  <tr>
    <td>aes-192-gcm</td>
    <td>24</td>
    <td>12</td>
    <td>16</td>
  </tr>
  <tr>
    <td>aes-256-gcm</td>
    <td>32</td>
    <td>12</td>
    <td>16</td>
  </tr>
  <tr>
    <td>chacha20-poly1305</td>
    <td>32</td>
    <td>12</td>
    <td>16</td>
  </tr>
  <tr>
    <td>chacha20-ietf-poly1305</td>
    <td>32</td>
    <td>12</td>
    <td>16</td>
  </tr>
  <tr>
    <td>xchacha20-ietf-poly1305</td>
    <td>32</td>
    <td>24</td>
    <td>16</td>
  </tr>
</table>

**Comment by wongsyrone** [comment]
All you need is listed here: https://github.com/shadowsocks/shadowsocks-org/issues/30#issuecomment-273693878

**Comment by madeye** [comment]
OK. Let's keep using current numbers from @wongsyrone.

**Comment by wongsyrone** [comment]
Since the non-ietf version of chacha20 is already released, the author will not remove that. Otherwise, he will break compatibility.

**Comment by wongsyrone** [comment]
If you really care about nonce size, recommend ppl to use the ietf version.

**Comment by riobard** [comment]
No, we have to be very careful here. 

The standard 12-byte nonce value is recommended under a vastly different assumption that key rotation is common (as in TLS), where as in Shadowsocks it's rare to change keys.

The proper guidance to choose nonce is here https://tools.ietf.org/html/rfc5116#section-3.1

**Comment by riobard** [comment]
Basically, the recommended way to deal with nonce and long-term keys is 

>  If a secret key will be used for a long period of time, e.g., across
   multiple reboots, then the nonce will need to be stored in non-
   volatile memory. 

Obviously we don't want to do that for simplicity reasons. Random nonces are easier to manage, but then it's basically deciding how much you'd like Birthday Paradox to hurt.

**Comment by riobard** [comment]
By the way, default nonce size for chacha20-poly1305 is actually only 64 bits (8 bytes)… which is definitely going to be a problem in our case.

**Comment by wongsyrone** [comment]
I want to remind everyone about https://gist.github.com/clowwindy/5947691; the AEAD design is not supposed to deal all security issues, it can be considered as a little enhancement to the previous OTA with sane authentication to against server-side behavior based attacks. The goal is to protect server's identity.
The attack is already elaborated by Noisyfox.

Security should never be placed in the first place. It is not the intention of Shadowsocks.
As said by clowwindy, it is a tool to help you bypass firewalls.

**Comment by Mygod** [comment]
Correct but sadly a lot of times bypassing firewalls requires security, for example the reason why we need to use encryption at all. In this case, I think we should make the nonce size sufficiently large so that in the foreseeable future the firewall won't have enough computational power to take advantage of this.

**Comment by xseven007** [comment]
This is a lever of covert and security. In the future, we should let the firewalls can not detect what kind of our connect is.

**Comment by riobard** [comment]
Again, we need to be very careful here to avoid future issues like what caused this redesign.

Nonce reuse allows attackers to generate more messages with the same nonce. Given that the critical Shadowsocks request header comes right after the nonce, it will potentially allow similar attacks to identify the server. 

Even if we stick with the default nonce size, I'd like to add the ability to increase it should we ever need to do so. So, making it into a command line parameter?

**Comment by riobard** [comment]
Once AES-SIV becomes widespread, this nonce reuse issue will be gone for good. There's still hope! ^_^

**Comment by xseven007** [comment]
Once the new version has setuped in the server, The legacy client will became useless. So shall we do something to let users know this change? Like a log on the screen on server "The Previous OTA function has been deprecated please fix the config file" ?

**Comment by riobard** [comment]
Probably it's easier to just call this Shadowsocks v2 and make a clean break.

**Comment by Mygod** [comment]
@wongsyrone And for your case, SSL provides a different aspect of security which is certification. As for Shadowsocks, being indistinguishable from random noise (for whoever doesn't hold the knowledge of the preshared secret) should be the topmost priority as any kind of pattern can be used to identify the traffic, even if it's not very obvious, like the plaintext length in the original OTA protocol. We don't need certification and having certification also hurts our randomness.

@riobard [RFC5297](https://tools.ietf.org/html/rfc5297#section-1.3.2):

> SIV provides a level of resistance to nonce reuse and misuse... If, however, the
> nonce is reused, authenticity is retained and confidentiality is only
> compromised to the extent that an attacker can determine that the
> same plaintext (and same associated data) was protected with the same
> nonce and key...

A "level of resistance" isn't exactly the same as "gone for good".

Although addition of AEAD makes the codebase a little bit messy but the user interface doesn't get any messier: to a normal user, there are just more options to the `method` parameter. It's good to keep the compatibility for now. Reasons:

1. Stream cipher without authentication is susceptible to breakwa11's detection but there seems to be no  clue (or is it?) that GFW is currently deploying it as that would be pretty obvious by reading log on server side; We should drop the support for OTA however because GFW is able to detect OTA without the endpoints knowing it.
2. Stream cipher without authentication is faster; (but [by how much?](https://github.com/shadowsocks/shadowsocks-libev/issues/1173))

I think we should however add a notice to shadowsocks server's log to educate the server master of the downsides of stream cipher if they are still using them. If they found in their log unknown atyp entries, they should migrate to AEAD asap. @madeye 

@xseven007 It's not possible to display something on a legacy client without updating them. And supposedly if someone has upgraded to the latest server, they would probably have changed their cipher to AEAD and it's impossible to communicate with the old cipher anyways. Publishing notices isn't Shadowsocks's responsibility.

**Comment by riobard** [comment]
SIV is nonce misuse-resistant, so shorter randomly chosen nonces don't hurt. As long as reusing nonce doesn't leak keys or allow fabrication of more messages using the same nonce, we'll be fine. 

Performance wise, on my Ivy Bridge 2.6 GHz Core i7 with AES-NI, AES-256-GCM gives me > 4Gbps throughput (simultaneous encrypting and decrypting) measured by iperf3.

**Comment by Mygod** [comment]
Yes I agree that we can use shorter nonce in that case.

Well that kind of benchmark doesn't really make a lot of sense. I was hoping to run some benchmark on my old computer, my cheap servers and my phone.

**Comment by xseven007** [comment]
@Mygod Yes , I understand. What I means is that, It is a important and difficult decision...Please be careful. And If we break the lever of covert and security. Something would be changed.

**Comment by madeye** [comment]
@riobard BTW, do you have time to add SIP004 to shadowsocks-go? Pull requests are welcome.

**Comment by wongsyrone** [comment]
According to some comments by clowwindy

> Or use authenticated encryption?
> sopium We need more than that. We need a stream cipher.

We can infer his goal is a kind of stream cipher with authentication embedded instead of separate a stream into chunks

> My goal is to design an authenticated symmetric-key encryption protocol

He also raises concerns about AE:

> There remains a problem:
> We have to **tell the decryptor about boundary information, which is dangerous**. An attacker can read the boundary information and know about the protocol.

refs:
https://github.com/shadowsocks/shadowsocks/issues/108#issuecomment-42102661
https://github.com/shadowsocks/shadowsocks/issues/64
https://github.com/shadowsocks/shadowsocks/issues/69

---

Based on these facts, I find a cipher might be related to clowwindy's goal.
https://link.springer.com/chapter/10.1007/978-3-642-28496-0_21

**Comment by wongsyrone** [comment]
As long as I know, clowwindy's purpose is to hide protocol correctly. If something breaks this principle, we have to give up the idea no matter how secure it is.

**Comment by wongsyrone** [comment]
At the same time, clowwindy also said

> If we use authenticated encryption, none of these attacks will work.

in https://github.com/shadowsocks/shadowsocks/issues/69#issuecomment-25622620.

I'm really confused.

**Comment by Mygod** [comment]
@wongsyrone Telling attacker/adversary where the boundaries are is the exact mistake we've made in OTA. AEAD fixed that. If we didn't make another mistake, AEAD should be as secure if not more secure than plain stream cipher and OTA in every way we want to achieve. The only valid downside I can think of is performance.

The current AEAD proposal is exactly designed for Shadowsocks. It's basically a stream cipher with an authentication tag appended at some locations.

**Comment by riobard** [comment]
@madeye I had actually looked at the shadowsocks-go code base last year. To be honest, it is in very bad shape. Maintaining it takes more work than a clean rewrite to be more idiomatic Go. I can open source my new implementation with AEAD crypto only, and gradually add back the original stream cipher suites for backward compatibility. If you're OK with it, we can retire shadowsocks-go. I believe it will lay a better foundation for future improvement.

**Comment by madeye** [comment]
@riobard Great! I think a new shadowsocks-go would be a good choice. When you're ready, let's add your implementation to the organization.

**Comment by riobard** [comment]
@madeye Your call, boss! :P 

I'll do some cleanup work and release mine soon. Then we can work on backward compatibility.

**Comment by wongsyrone** [comment]
We are indeed making mistakes again.

> A common mistake is to use the same salt in each hash. Either the salt is hard-coded into the program
> This is ineffective because if two users have the same password, they'll still have the same hash.

Just as I said before.

https://crackstation.net/hashing-security.htm [The WRONG Way: Short Salt & Salt Reuse]

**Comment by Mygod** [comment]
@wongsyrone Password hashing != key derivation. And in our case, we have random IV.

**Comment by wongsyrone** [comment]
It has nothing to do with the random IV.

**Comment by wongsyrone** [comment]
The current Argon2i way is a looking-secure-but-pointless way. I will not adopt this part until fixed.
Or you can kick me out.

**Comment by madeye** [comment]
I think the reason we choose Argon2 here is because we need a modern password hashing algorithm for our key derivation. No matter which key derivation we choose here, there is always a "salt" issue.

**Comment by madeye** [comment]
@wongsyrone What's your proposal here?

**Comment by wongsyrone** [comment]
Since you always have the same input on both client and server side, it will generate the same output.

There is no benefit from password hashing. The salt used in it is for the same user-password, we can get different hashes. But it is not suitable for Shadowsocks. If you get the different key, you cannot do decryption.

My proposal is a general hashing algorithm. You can keep the old MD5 way or a modern one.

**Comment by madeye** [comment]
So, the question is why generic hashing is better than our old key derivation approach (MD5)?

**Comment by wongsyrone** [comment]
It claims to be faster and as secure as SHA-3.
If there are no pieces of evidence, we can use the old way.

**Comment by Mygod** [comment]
MD5 is a generic hashing algorithm but collisions for MD5 has been found so I personally veto MD5 even if it does not seem to be a big issue here. I still prefer to getting rid of key derivation and requiring a random key like GPG.

**Comment by wongsyrone** [comment]
> collisions for MD5

This is also my concern.

**Comment by Mygod** [comment]
Also using same salt for password hashing isn't quite a big issue here either. We should still have security as good as using generic hashing.

**Comment by madeye** [comment]
I'm OK with any key derivation approach here. Actually I don't think it's really important to us.

**Comment by wongsyrone** [comment]
The same hashing result is required by Shadowsocks itself since we don't have the key negotiation.

> Actually I don't think it's really important to us.

Yes, we only need to reduce collisions.

**Comment by madeye** [comment]
> collisions for MD5

Any possible attack to this? Is it really important to shadowsocks?

**Comment by wongsyrone** [comment]
According to https://download.libsodium.org/libsodium/content/hashing/generic_hashing.html

> Unlike MD5, SHA-1 and SHA-256, this function is safe against hash length extension attacks.

But I have no idea on hash length extension attacks.

**Comment by Mygod** [comment]
@madeye Afaic, no.

@wongsyrone HMAC isn't prone to hash length extension attacks. We're fine here.

I still suggest to use random key to prevent server masters without cryptography knowledge screwing up and get rid of these possible concerns once and for all. Some server masters could even use a single int32_t as a password. In that case no matter what key derivation function we use the password is still weak.

**Comment by wongsyrone** [comment]
But how to let users input a long random string into Android app? It is not user-friendly.
We may add some advice for password.

**Comment by wongsyrone** [comment]
User input is not that random like hashing results.

**Comment by Mygod** [comment]
@madeye Even though I can't conceive how attackers can take advantage of MD5, MD5 is generally considered insecure and it would be weird to see MD5 in our claimed-to-be-secure protocol. :sweat: 

@wongsyrone Just like PGP, use base64. And we can use QR code for our case, I don't really manually input password on Android app anyways.

And that's the problem with user input. User input has relatively lower entropy which adversary can take advantage of.

**Comment by wongsyrone** [comment]
QR code still has its shortage. It is not a good idea to rely on it.

**Comment by Mygod** [comment]
What shortage? It's only used for configuration. But anyway at least you can rely on the underlying ss URI string.

**Comment by wongsyrone** [comment]
We should use SHA-1 at least IMO.

**Comment by wongsyrone** [comment]
As long as I know, xSocks and ShadowVPN all uses Blake2b.
I don't think it a big problem here.

https://github.com/lparam/xSocks/blob/master/src/crypto.c#L64
https://github.com/wongsyrone/ShadowVPN/blob/master/src/crypto.c#L42

**Comment by icodesign** [comment]
Sorry to interrupt. I found that we have changed the method to generate encrypt key. Does that mean old clients won't be able to access 3.0 servers?

**Comment by wongsyrone** [comment]
It is not changed for old stream methods. It is only applied to AEADs.

**Comment by xseven007** [comment]
@icodesign If you want to use the old client to access the 3.0 server, maybe you should disable the OTA on the client side.

**Comment by riobard** [comment]
+1 for long random keys. How that key is generated can be left to deployment. It'll simplify the code base as well.

**Comment by riobard** [comment]
Personally I just `xxd -p -l 16 /dev/urandom` to get 16-byte random key in base16.

**Comment by Mygod** [comment]
Current key derivation algorithm is also secure enough as long as the user knows how to correctly set a strong password. If you can be bothered to manually type your password in shadowsocks-android, it's an implication that your password isn't strong enough. :stuck_out_tongue:

As I've said, random key is more secure and can prevent uneducated server masters to screw up.

**Comment by wongsyrone** [comment]
Then you should do length check, although most of them share the same key length.

**Comment by Mygod** [comment]
@riobard We can use something like `gpg --gen-key` which uses user input, disk activity, etc. as random entropy source. (but that one is for generating key pairs)

@wongsyrone Correct.

**Comment by madeye** [comment]
OK, let's pick Blake2b as default key derivation for AEAD.

@Mygod This proposal looks interesting. I think it can become another SIP. It allows us to put long random keys in the URL and QR code, rather than deriving keys from password.

**Comment by Contextualist** [comment]
@riobard For shadowsocks-go reconstruction, have you checked @lixin9311 's  works(https://github.com/shadowsocks/shadowsocks-go/issues/192#issuecomment-274905015)? It seems that he/she is working on reconstruction. Maybe you two could collaborate.

**Comment by Mygod** [comment]
@madeye Argon2i also looks fine to me. Well Blake2b is probably faster since it's a generic hashing algorithm.

**Comment by wongsyrone** [comment]
Argon2i is indeed a good password hashing algo, but you don't benefit from it. You did all memory-hard computation, get poor performance, be low-end-box-unfriendly, and have the same result as general hashing algo.

What we need is just a long string to become key for encryption.

**Comment by riobard** [comment]
@wongsyrone Not really. Even if using the same salt (equivalent to no salt), slower password hashing still gets us the protection from rainbow tables because of much higher computation cost to brute force a password. 

If we insist on using password instead of random keys, at least use proper password hashing.

**Comment by wongsyrone** [comment]
Does this tradeoff really matter? I want to make it low-end-box-friendly.

**Comment by riobard** [comment]
It depends. As mentioned by @madeye above, current AEAD cipher suites are already quite taxing and thus not very suitable for low end routers. Maybe we should just let it go. Low end routers should stick to the original stream ciphers.

**Comment by Mygod** [comment]
@wongsyrone In that case you should be happy about SIP006 since it actually makes it much more friendly to low end boxes.

**Comment by wongsyrone** [comment]
My proposal just adds little security to hashing, but keep on the track of the original Shadowsocks way.

**Comment by wongsyrone** [comment]
If you want to live shorter, you can try SIP006.
We get many issues asking how to compile new shadowsocks-libev these days, only with little change to configuration and dependencies. I can't imagine what will happen if you implement SIP006.

**Comment by Mygod** [comment]
The original way isn't necessarily the best way. And Shadowsocks isn't convenient to set up since the beginning. The sooner we put it out there the sooner we will see people writing tutorials how to properly set up the server. And shadowsocks-libev is currently just a pre-release ATM. It might as well be a good time to do that.

**Comment by madeye** [comment]
It looks a little bit off-topic here.

@wongsyrone If we choose another password hashing, e.g. PBKDF2, could you accept that?

**Comment by wongsyrone** [comment]
As I said before, there is no need to use password hashing.

**Comment by madeye** [comment]
So, @riobard If we use Blake2b here and keep pushing SIP006, does it work for you?

**Comment by riobard** [comment]
Maybe we should discuss this under https://github.com/shadowsocks/shadowsocks-org/issues/35

@madeye I don't really mind. It's just kinda pointless and silly to switch to another generic hashing.

**Comment by wongsyrone** [comment]
You win.
I quit.

**Comment by Mygod** [comment]
Well I'm leaving this decision up to @madeye. I think I have listed all my points.

**Comment by madeye** [comment]
@wongsyrone 😓 

@Mygod 😓 💦

**Comment by riobard** [comment]
@wongsyrone Come on, buddy. Don't be childish.

**Comment by wongsyrone** [comment]
At the very beginning, I am keeping describe everything. It results in that way. I have no idea and don't want to take part in it anymore.

Do what you like and good luck.

**Comment by Mygod** [comment]
@wongsyrone Well you have a few solid points but overall I think the pros (of SIP006) outweigh the cons IMO. But hey that's just my opinion. Nobody says you're wrong.

**Comment by riobard** [comment]
I don't even know why @wongsyrone is pissed off here… Anyway, please continue the discussion regarding KDF on SIP006. This thread is getting really off topic now.

**Comment by xseven007** [comment]
@riobard The more security the shadowsocks is the more covert we lose. https://twitter.com/clowwindy/status/638319982074421248

**Comment by riobard** [comment]
@xseven007 What are you talking about? He's referring to something completely different.

**Comment by Mygod** [comment]
@xseven007 You should get a hold of what we're doing before writing random comments. Why don't you use minimal security by transmitting raw text to get maximum covert as how you've put it?

**Comment by xseven007** [comment]
@Mygod @riobard Sorry about that. I have not see all about what we're doing in this issue. Then I will learn this issue.

**Comment by Mygod** [comment]
For others who may have the same or similar concerns like @xseven007 did, we are not blindly throwing anything that can improve our security, like session keys. We are only adding things that improves security without hurting our main goal, which in turn helps us to achieve our main goal.

**Comment by echoowall** [comment]
我这种四级还没过的，就随便发一条刷一下存在感，反正我也看不懂。匿了么么哒

**Comment by mindcat** [comment]
Somehow we are using English makes peoples like @echoowall which is didn't understand what we are talking about =X

Sorry if I am offending.

**Comment by mindcat** [comment]
I could be wrong, maybe its not worth it. Anyway I will checking the shadowsocks new protocols.

**Comment by lenovoy450** [comment]
![photo_2017-02-05_14-42-21](https://cloud.githubusercontent.com/assets/24227226/22624275/b2c4d690-ebb1-11e6-89a8-cfa7b6cdef2d.jpg)

**Comment by riobard** [comment]
Please leave aside political argument here. No time for such useless shaming and blaming.

**Comment by Mygod** [comment]
I've read some people's opinions against this proposal but only valid points are performance. As shown in [benchmark](https://github.com/shadowsocks/shadowsocks-libev/issues/1173#issuecomment-277184307) AEAD doesn't add much overhead. If you want maximum performance you can still use stream cipher. In my opinion detection via packet length, timing, etc. is something SIP003 plugins should consider. The responsibility of Shadowsocks is to produce high entropy ciphertext.

@breakwa11 If you think there's anything wrong/unnecessary with this proposal (or other proposals) or this proposal is going in the wrong direction, feel free to add your comments, preferably with some explanations. The same also applies to everyone else.

**Comment by printempw** [comment]
Hello I want to know some details about how GFW able to detect shadowsocks servers with OTA enabled. 
The origginal shadowsocks protocol could be detected by servers's reaction when receiving packet with modified iv, which could be tampered with easily since original shadowsocks uses stream cipher. (If I understood it correctly) But according to your discussions above, it seems that GFW could detect servers with OTA enabled even with endpoints knowing it.

How do the detection works? I've read the documentation of OTA, is there any problem about the packet length?

**Comment by wongsyrone** [comment]
因为之前有人说英文看不懂，而且我怕表达不清，这里用中文。

这个draft是我最先贴上的。当时的设计这个的理由是：看到文章说加密和认证的组合有三种分别是MtE，EtM和E-and-M，其中只有EtM是加密业界推荐的模式，其他的两种据说都有漏洞，于是要解决这个，把TAG移动到密文。之后因为切分stream成chunk，需要有长度信息用来算offset，加了len字段用来搞定可变部分。为了保护len不受篡改，经过加密之后加了个对应的tag。

但是后来很多人给了中肯的建议，尤其是 @riobard 提到对于nonce误用导致的致命缺陷。这个是设计的时候没有考虑到的。一些分析可以见 https://github.com/shadowsocks/shadowsocks-org/issues/36

TLS里面有可以解决引入问题的方法，但是那些受限于SS原协议，不便加入，否则影响用来翻墙的目的。
我要说的就这么多。

**Comment by v3aqb** [comment]
其实len改过之后MAC肯定是过不了的

Syrone Wong <notifications@github.com>于2017年2月5日周日 19:26写道：

**Comment by wongsyrone** [comment]
见上文Noisyfox对于服务器行为的分析

**Comment by breakwa11** [comment]
@Mygod 我英文非常差的，我看不懂你们在讨论啥，所以也没有什么意见好给的，不管设计出来什么样，到时候去实验测试就知道结果了，多一种想法没什么不好的

对于前面图里那位叫做 Si Chen 的兄台可能他看不懂英文，以为写了我的名字就是我在推，这里我要说明一下

对长度加校验的这个做法，我最早是2016年9月在自己的协议中实现的，后来把这样做的原因和 @Noisyfox 交流，他理解了，再后来， @wongsyrone 来找我和 @Noisyfox ，他说他想自己设计协议，通过交流，我否决了他早期的方案后，结果就是前面你们看到的那个，所以这个AEAD协议设计者是 @wongsyrone ，他部分采用了我的做法，所以引用了我的名字。至于推进AEAD，更和我没关系，我不是这个SIP004的发起人，更全程没有参与这个issue的讨论。

所以，推进AEAD和我完全无关，否决OTA这个决定也和我完全无关，我也不是SS项目的决策参与者，到底是谁推进和谁否决，请那位兄台认真读完本issue，这里有答案。所以呢，你们啊，就总想搞个大新闻，明明我全程没参与决策，非要强行说炒作、带节奏、歪楼，无视事实的做法简直low爆了，这种做法简直TOO YOUNG TOO SIMPLE，Sometimes NAIVE。你们这些西方媒体啊，还是要提升自己的姿势水平，不要见风就是雨，将来要是宣传上出了偏差，你可是要负责的。

**Comment by librehat** [comment]
@Mygod I'm very late for the discussion. But I saw from your previous post that GFW can detect shadowsocks connection with OTA enabled. Would you elaborate a bit on attacks on OTA-enabled shadowsocks?

**Comment by printempw** [comment]
@librehat The structure of an OTA-enabled chunk (decrypted) of shadowsocks TCP relay is like this:
```
+----------+-----------+----------+----
| DATA.LEN | HMAC-SHA1 |   DATA   | ...
+----------+-----------+----------+----
|     2    |     10    | Variable | ...
+----------+-----------+----------+----
```

DATA.LEN is a 16-bit big-endian integer indicating the length of DATA.

Since OTA is based on **stream cipher**, where each plaintext digit is encrypted one at a time with the corresponding digit of the keystream, to give a digit of the ciphertext stream, GFW could tampered with the ciphertext with the same method [mentioned](https://github.com/breakwa11/shadowsocks-rss/issues/38) by @breakwa11.

Once ciphertext is modified, the decrypted `DATA.LEN` is **not** as-is anymore. 

e.g. If the `DATA.LEN` is modified to be longger then the original, the shadowsocks server will wait until the **really non-existent** bits arrived, which could cause attacks based on server behavior. Unlike the detection for the original shadowsocks protocol, OTA-enabled shadowsocks servers could be detected even **without endpoints knowing it**!

Be free to criticize if I havn't understood it correctly :)

**Comment by wongsyrone** [comment]
既然这样可以改成

designed by wongsyrone with help from breakwa11 and Noisyfox

**Comment by Mygod** [comment]
@madeye This has just passed my mind: did we use a certain bit flag for client/server packet? If not, we're literally creating nonce collisions all the time.

**Comment by riobard** [comment]
@Mygod No. The connection is encrypted independently on each direction (two keys).

**Comment by madeye** [comment]
@Mygod @riobard I think two independent nonces?

**Comment by riobard** [comment]
@madeye Correct. Independent nonces. Although the discussions in #36 and #40 still apply.

**Comment by Mygod** [comment]
@madeye @riobard Oh okay.

**Comment by shinku721** [comment]
I firmly support this enhancement personally, but I must point out that it cannot help shadowsocks to get rid of the risk completely.

In the new proposal, as I assumed, length with its tag is checked firstly, then the payload and its tag.

Now we can do a detection:
1. record the stream of a certain connection
2. change some byte and detect the length of further bytes we send that causes the server to close the connection
3. now change the next byte beyond the length in step 2, and repeat step 2.
4. finally we break up the stream into chunks

Since the length of (Length_Tag + Length) is fixed, we can see this pattern appear (maybe repeatedly if more than one chunk in the uploading stream) in those chunks. Then we are almost sure that the server is a shadowsocks.

It is possible to delay the time of disconnection, 

Actually I don't think the security of OTA is compromised, though it is MtE. HMAC without padded block cipher mode (like CBC) has no obvious vulnerabilities currently. This is somehow a replay attack (the same request header passes the check and the data is performed). I may propose a TOTP scheme later to defend replay attacks.

**Comment by Mygod** [comment]
To defend against this kind of attack, we should probably drop malicious requests instead of closing the connection.

The biggest problem with OTA is that it's length in transmitted in plain text, not MtE.

**Comment by shinku721** [comment]
@Mygod Really? I have implemented shadowsocks myself for serval times and I always encrypted the length field together with MAC and payload, and it seems to be compatible with libev versions.

**Comment by Mygod** [comment]
Hmmmm I guess I misunderstood the documentation...

**Comment by printempw** [comment]
The `DATA.LEN` **is** transmitted as ciphertext, actually.

But the biggest problem with OTA is that adversaries can build evil packets and do server behavior based attacks since `DATA.LEN` is not authenticated in OTA.

Some details about the attack: https://blessing.studio/why-do-shadowsocks-deprecate-ota/

**Comment by madeye** [comment]
@shinku721 One interesting thing is most of inner traffic of shadowsocks is TLS. It means even with stream ciphers, we would see chunks in shadowsocks traffic.

Is it a characteristic? Yes, it is. But is it really useful for attacking? No, I don't think so.

**Comment by shinku721** [comment]
@madeye Chunks are not the pattern, but the length of the chunks is.
However, I agree with you that this cannot be a practical attack. I mean, it has some characters similar to the `weakness of OTA` mentioned above, that it has to replay the request header. Both attacks are currently in theory and no evidence can prove that the GFW has already adopted any of them.

The reason why I support AEAD ciphers is just that AEAD ciphers are widely used in current practice.
IMO there is no urgent need to deprecate OTA, but we must stop using the original protocol without forcing OTA.

**Comment by riobard** [comment]
Please note that SIP004 is superseded by SIP007 (#42).

**Comment by tigaly** [comment]
从梅林被老司机领到这里，我就想问问大家，有什么决定性的代码或者证据证明OTA是先天不足必须给钦定的AEAD让路么？如果有，请联系我，我提供个C段地址，求扫描

**Comment by hellofwy** [comment]
@tigaly 
可参考：https://blessing.studio/why-do-shadowsocks-deprecate-ota/，当然我并不是完全认同文章的内容。
就我这段时间对他们讨论的观察，决定用 AEAD 主要对安全的的考虑，并不是防识别。
因为现在也引入了插件功能，可以通过插件实现协议混淆防检测，或实现协议伪装欺骗运营商 QoS 获取加速，或者其他功能。

**Comment by tigaly** [comment]
@hellofwy 这篇文章针对OTA不就是那个很容易fix的漏洞么（通篇吹破哇）？还有什么直接证据呢？那么，出于安全级别的考虑有没有什么别的漏洞无解呢？

**Comment by hellofwy** [comment]
@tigaly 
就我目前知道的，并没什么直接证据。
安全这东西无止境，建议关注 TLS 协议最新相关研究，毕竟这个是世界范围内广泛使用和关注的。
作为普通使用者，在意安全的时候，使用 HTTPS 就好。不要太依赖底层 ss 协议。

**Comment by tigaly** [comment]
@hellofwy 我并不是在意安全问题，我只是用梅林的时候发现了OTA更新被直接干掉了，才想来学习下，不过这种漏洞解决的方式真奇怪哈，要是哪天证明AEAD也有问题？是不是要搞个DEAD协议出来？

**Comment by v3aqb** [comment]
😁

Tigaly <notifications@github.com>于2017年2月11日周六 17:15写道：

**Comment by breakwa11** [comment]
@tigaly 既然不在意安全问题，那你直接用最早的版本的协议就好了，用OTA干什么？OTA本质就是增强安全为目的设计的

**Comment by hellofwy** [comment]
@tigaly 
想太多。
OTA 一直也只是标为实验性功能。
AEAD 是信息安全领域的已有名词，并不是这里的人自己创造的。

**Comment by tigaly** [comment]
@breakwa11 那你去掉OTA就是为了安全？有没有证据可以证明呢？我并不在意安全问题，并不是我心大到完全不在意安全，只要gfw探测不到就行了

@hellofwy 我当然知道，只是开玩笑的说法，对于大家这种严谨的做法，我真是大开眼界

**Comment by breakwa11** [comment]
@tigaly 什么叫做“我去掉OTA”？？？？？？？？？？

**Comment by tigaly** [comment]
@breakwa11 除了你还有谁一直在鼓吹OTA有不可修复的漏洞？

**Comment by breakwa11** [comment]
所以我也可以说：windows和linux有一大堆漏洞，然后修复漏洞的人就是我了，是这种逻辑吧，太好了，成百上千的都是我弄的

**Comment by tigaly** [comment]
@breakwa11 你这手驴打滚偷换概念玩的6，我也懒得跟你说，OTA和AEAD这两个功能的更替，有多少你在里面带节奏，明眼人一看就知道，不是你想推脱就推脱的

**Comment by breakwa11** [comment]
那不然呢？SS的代码我一行没动，在一行没动的情况下就成了是我改的？我不动一根手指但几千代码都是我改的，哇，难不成你说madeye的号被盗了？

**Comment by RobertYan** [comment]
@tigaly @breakwa11 Please stop posting any irrelevant comments here.

**Comment by tigaly** [comment]
@breakwa11 有种叫教唆犯你不会没听过吧

**Comment by hellofwy** [comment]
@tigaly 
这里讨论关于人的问题不好。
毕竟这里说最好都是软件的问题。
针对人的在其他地方讨论就好。

**Comment by tigaly** [comment]
@hellofwy 是啊，本来说的好好的，不知道哪来的人跳出来一阵节奏和歪楼，
另外我翻了下好多关于OTA的讨论都完全被人带节奏，TLS和SS的用途被混为一谈，我就不说了，反正我也不怎么用ss，你们大家慢慢加油吧

**Comment by qinyuhang** [comment]
@Tigaly  abandon of OTA is discussed，cannot undo. Besides one saying says U
can U up， talk is cheap show  your code <notifications@github.com>于2017年2月11日
周六17:52写道：

**Comment by tigaly** [comment]
@qiuyuzhou 我进这个讨论组第一句话就说了，谁能针对OTA探测，求扫描，我提供一个C段地址，只是其他人一直歪楼带节奏才变成这样

**Comment by qinyuhang** [comment]
U can privately send ur IP to breakwa11，instead of start a fight here.
Tigaly <notifications@github.com>于2017年2月11日 周六18:03写道：

**Comment by tigaly** [comment]
@qiuyuzhou 如果她敢接就不会一直打嘴仗了，从sadoneli一直扯到这里来

**Comment by qinyuhang** [comment]
She already post her pubkey，just gpg encrypt it and post your encrypted IP
here
Tigaly <notifications@github.com>于2017年2月11日 周六18:08写道：

**Comment by madeye** [comment]
Locked as off-topic.

**Comment by qiuyuzhou** [comment]
@tigaly 你@错人了。。。
